### DISCUSSION TITLE
new update to emulatorjs.com

### DISCUSSION THREAD

**Discussion by ethanaobrien** [EmulatorJS/EmulatorJS#223] [General]
Heyo yall. I enabled discussions so we have a place to talk / ask questions without needing to open an issue.

<br>

I recently found out, the emulatorjs.com website recently pushed put a new update.

At first I was like yayyyyy but when looking at the code I was like noooo.

Announcing these things:
1. I am no longer planning on using any new source code from the emulatorjs website. We will build onto what we have.
2. Yeah thats basically it

So why are we no longer using it from their website? Well the answer is simple (not really). Heres a list of why
1. The new emulator code is much much more obfuscated then before (much much harder to read and de-scramble)
2. After lots and lots of work, I managed to deobfuscate the new `emulator.js` and `loader.js` files. (I have the attached files at the bottom)
3. They implemented a security mechanism that requires the wasm file to communicate with the server and this cannot be bypassed. The wasm will not run without the key from the server (the key is different every time, I already checked this)
4. I dont want to have to re-write all the changes we all made
5. This is the biggest reason, from what I have found the beta cores are already so much faster (and use less cpu) then the cores they use. I do not think its worth the work since we are so close to being done with beta (yeah I'm stuck at the hardest part but being able to really play the game is what matters)

As promised (it wouldnt let me upload them straight to github, even as a zip. Weird):
https://drive.google.com/drive/folders/1s0FeGe2SdhWDmgpg3jCZPV9YjtdLNks5?usp=sharing

I will not be sharing the new deobfuscating files I wrote, as I think when I released them last time Im pretty sure the emulatorjs people used them to test to make sure they would not work on the new files. So sorry guys but its for the better.

When you're looking through the files and you see something that might make a cool feature, make an issue! I'd love to try to implement some new stuff!

For example I plan to (in the `loader.js` file) cache the `emu-min.js` file through localstorage, which is something new they implemented which will allow faster load times

Question for y'all, do yall think we should stick with the current svg button icons or use the new updated ones?

Please dont be afraid to say something! As long as it is appropriate or mean! In someones life youre making an influence with your little comment/idea.

**Comment by Ramaerel** [comment]
I do like the new design layout they have, and the inclusion of a d-pad is really nice. Although, the l and r buttons are place weirdly...

Are there any other technical changes that you can see

**Reply by ethanaobrien** [reply]
Not that I've found so far. The first thing I did see when I saw their new emulator was this really annoying bug with the settings button that was actually really irritating.

Do you think I should make an option to change between the new virtual gamepad they have and the old one? I dont think It would be that hard (I do also plan to try to update the code for the webkit gamepad api)

**Reply by Ramaerel** [reply]
I did notice that the settings menu pops up immediately on load on the new update, which is kind of a pain. I don't know if that's what you're referring to but yeah

The option to swap between the two gamepads would be good. I know it's a bit of an out there question but is there anyway modular gamepads would be possible at some? Like, files the users can edit to change the gamepad for the site per system

**Reply by ethanaobrien** [reply]
There is a way the user can edit the gamepad, (its not very easy but yeah)

In the `emu-main.js` file, search for `toggleVirtualGamepad`. It should be the first result. This is where you can edit the code

**Reply by felipermfalcao** [reply]
Hi! I'm new here and I'm really enjoying the emulator. One question: Is it already possible to switch between the old and the new virtual gamepad? I use it on my cell phone and the analog directional pad is sometimes bad, I prefer the buttons.

**Reply by felipermfalcao** [reply]
There is and using the cell phone too, how do I enable Quick Save and Quick Load?

**Reply by ethanaobrien** [reply]
The new virtual game pad is not yet implemented

on a phone, there isn’t exactly quick save and quick load, go into settings and change the save state location to your browser, then when you save/load it’ll read the file from your browser. If you want to download your state, just change this back

**Reply by felipermfalcao** [reply]
Many thanks for the reply. I tried to emulate the DS mario RPG 3, but it was slow. Is there a site with compatible games? I'm using a POCO X3 from Xiaomi.

**Comment by allancoding** [comment]
This emulator has a lot of future what we have now is the hardest part. Please list any features that you would like ideas are what we need because you are the people using it! It's not like you see a request feature button on there website.😅 I plan this summer to make the beta into the main cores. 

And I think that we should do a vote on the new icons. And have the option for people to suggest different ones.

**Reply by ethanaobrien** [reply]
Just so you know, #151 contains the todo list

**Reply by allancoding** [reply]
Ok, thank you for keeping this repo updated!

**Reply by ethanaobrien** [reply]
Heres some helpful information that will help you with building the beta cores

1. Ive only ever gotten it to compile in ubuntu (I use ubuntu 20.04 in WSL)
2. All the shader commands are located in `/menu/menu_driver.c`
3. The default input driver (for webassembly) is `/input/drivers/rwebinput_input.c`. This is by far the most difficult task (Ive tried and havent made much progress, maybe youll be able to do it) and is also one of the most important. We will need to basically re-write the driver - yayyy (boooo)
4. I havent done any looking into catching when the rom fails to load. Sorry
5. And with the cleaning up the menu, a lot of this is going to be settings in the `Makefile.emscripten`, but these arent always defined correctly in the files so thats going to be a pain to go through all the errors and make sure they are in an if statement checking if the menu option is on


You dont need to bother on trying to compile the other cores if you dont want to, Ill do that part.

The repo to the emulatorjs retroarch source: https://github.com/ethanaobrien/RetroArch


Also, I went ahead and added you as a collaborator (to this main repository and the retroarch one). Please do everything through pull requests and wait for me to give the ok to merge. I would like to review the code first

**Reply by allancoding** [reply]
Ok I will just jump into ubuntu and get that done. And yes I will let you merge. And for the CORS errors I am seeing it is just with the new update because the older version works but the new one does not. I guess it will be solved in beta anyways! 
I will email you if I get stuck.

**Reply by ethanaobrien** [reply]
Awesome, thank you.

And with the cors errors, I'm 80% sure I fixed them already. Could you try on another device to see if the broken emulatorjs file is just being cached?

**Reply by allancoding** [reply]
I tried it on other computers and all it does is loads a black screen. But the beta loads.

**Reply by Ramaerel** [reply]
I have the same issue still too, it seems to load up fine but just sits on a black screen. No console errors or anything

**Reply by allancoding** [reply]
And it is just on the new version because my older version that I have works fine.

The issue is opened here: #233

**Comment by fallenking1064** [comment]
how do i get roms to work for this?

**Comment by ethanaobrien** [comment]
I was doing some browsing on the web and I found another thing, the new emulatorjs update seems to change the UI when the screen is compressed. (see [this](http://nesninja.com/game/nes/super-mario-bros#.YoB8k47MK00))

![Screenshot 2022-05-14 11 09 07 PM](https://user-images.githubusercontent.com/77750390/168456804-8ff401c1-40ff-400e-8d55-53c8bbdb7a8e.png)

If you look there is a button in the top right to toggle the visibility. I personally dont like it (it seems bulky and I just dont find it organized) but what do you guys think? Is this a feature you guys would want?

**Reply by Ramaerel** [reply]
The button in the top corner as opposed to where the menu button is seems like a good idea, but the menu ui itself is much nicer the way you have it now. I have no idea why they thought changing it was a good idea

**Reply by ethanaobrien** [reply]
I had the same thought actually, of replacing the menu button with the three lines but not changing it to that.

**Reply by allancoding** [reply]
I was looking at the new emulator and I found new shaders and audio quality and color palet should we add these in beta I think we can easily do more shaders. And I agree with the menu button idea.

**Reply by xq4894253** [reply]
I like this Menu UI

**Comment by Ducci9129** [comment]
did the emulator originally run offline, or did it always require a connection?

**Reply by ethanaobrien** [reply]
The emulator always has and always will require a connection, but if users want to be able to use offline then the site owners can look into using service workers and cache every request, so you can use the emulator 100% offline. My emulator button ( https://github.com/ethanaobrien/RetroArch ) repository has an example of this (the page will load when offline. Although the emulator button uses indexeddb to cache the files and then the EJS_paths variable to set them

Let me know if you have any Moreno questions

**Comment by DaBeybladeMaster** [comment]
save state button stopped working for me.
